mono/mini/interp/interp.c

```diff
@@ -6535,42 +6505,40 @@ interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClauseArgs
 			ip ++;
 			MINT_IN_BREAK;
 		MINT_IN_CASE(MINT_ENDFINALLY) {
-			gboolean pending_abort = mono_threads_end_abort_protected_block ();
-			ip ++;
-
-			// After mono_threads_end_abort_protected_block to conserve stack.
-			const int clause_index = *ip;
-
-			// clause_args stores the clause args only for the first frame that
-			// we started executing in interp_exec_method. If we are exiting the
-			// current frame at this finally clause, we need to make sure that
-			// this is the first frame invoked with interp_exec_method.
-			if (clause_args && clause_args->exec_frame == frame && clause_index == clause_args->exit_clause)
-				goto exit_clause;
+			mono_threads_end_abort_protected_block ();
+			guint16 clause_index = *(ip + 1);
 
 			// endfinally empties the stack
 			vt_sp = (guchar*)frame->stack + frame->imethod->total_locals_size;
 			sp = (stackval*)(vt_sp + frame->imethod->vt_stack_size);
 
-			if (finally_ips) {
-				ip = (const guint16*)finally_ips->data;
-				finally_ips = g_slist_remove (finally_ips, ip);
-				/* Throw abort after the last finally block to avoid confusing EH */
-				if (pending_abort && !finally_ips)
-					EXCEPTION_CHECKPOINT;
-				// goto main_loop instead of MINT_IN_DISPATCH helps the compiler and therefore conserves stack.
-				// This is a slow/rare path and conserving stack is preferred over its performance otherwise.
-				goto main_loop;
+			guint16 *ret_ip = *(guint16**)(locals + frame->imethod->clause_data_offsets [clause_index]);
+			if (!ret_ip) {
+				// this clause was called from EH, return to eh
+				g_assert (clause_args && clause_args->exec_frame == frame);
+				goto exit_clause;
 			}
-			ves_abort();
+			ip = ret_ip;
+			// FIXME pending abort ?
```

Comment:
I'm trying to convince myself that it's benign not to handle the pending abort here.  I guess the old behavior was to do an exception checkpoint once we're out of the outermost dynamic `endfinally`.  I think in normal control flow this just delays the interrupt check until some later instruction.

The thing I'm unsure about is when the finally clause is called by the exception handler.  In that case we'll unwind this `finally` and potentially have an unhandled exn instead of a thread abort.  I'm not sure the unwinder does any interrupt check on its own.

Reply:
I'll have to fix this. I just wanted to see the state of CI sooner.

Deep inside, probably I just wanted to not deal with aborts anymore, especially since the future of interp is on the .net core profile.